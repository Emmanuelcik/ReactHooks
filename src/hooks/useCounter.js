import {useState} from 'react';


export const useCounter = (initialState = 1) => {
    
    const [counter, setstate] = useState(initialState); //0 

    const increment = () => {
        setstate(counter + 1);
    }
    const decrement = () => {
        setstate(counter - 1);
    }
    const reset = () => {
        setstate(initialState);
    }

    return [
        counter,
        increment,
        decrement,
        reset,
    ]
}
