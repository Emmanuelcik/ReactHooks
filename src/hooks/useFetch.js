import React, {useState, useEffect, useRef} from 'react'
import "../components/02-useEffect/effects.css";

export const useFetch = (url) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        isMounted.current = true;
        return () => {
            isMounted.current = false;
        }

    }, []);

    useEffect(() => { 

        setState({data:null, loading: true, error: null});
        
        fetch(url)
            .then((res) => res.json() )
            .then( data => {

                if(isMounted.current) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }else{
                    console.log("set state no se llamÃ³");
                }
            });

        // return () => {
        //     cleanup
        // }
    }, [url]);

    return state;

}
