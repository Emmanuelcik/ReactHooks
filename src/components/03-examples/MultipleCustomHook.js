import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'

export const MultipleCustomHook = () => {

    const {counter, increment } = useCounter(1);

    const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;

    const state = useFetch(url);
    
    const {loading, data} = state;

    // const {author, quote} = data[0] | undefined;
    const {author, quote} = !!data && data[0];

    console.log(author);
    console.log(quote);
    return (
        <div>
            <h1>BreakingBad Quotes</h1>
            <hr/>

            {
                loading ?
                    (
                        <div className="alert alert-info text-center">
                            loading...
                        </div>
                    ) 
                : 
                    (
                        <blockquote className="blockquote text-right">
                            <p className='mb-2'>
                                {quote}
                            </p>

                            <footer className="blockquote-footer">
                                {author}
                            </footer>
                        </blockquote>
                    ) 
            }
        
            <button 
                className="btn btn-success"
                onClick={increment}
            > 
                Siguiente frase 
            </button>

        </div>
    )
}
