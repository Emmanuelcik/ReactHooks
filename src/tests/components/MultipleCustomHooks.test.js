import React from 'react';
import {shallow} from 'enzyme';
import { MultipleCustomHook } from '../../components/03-examples/MultipleCustomHook';
import { useFetch } from '../../hooks/useFetch';

jest.mock('../../hooks/useFetch');

describe("Testing the component MultipleCustomHooks", () =>{

    test("should be correctly mounted", () => {
        const wrapper = shallow( <MultipleCustomHook />);

        useFetch.mockReturnValue({
            data: null,
            loading: true,
        });


        expect(wrapper).toMatchSnapshot();
    });

    test("should get all info", () => {

        useFetch.mockReturnValue({
            data: [{
                author: "Emm",
                quote: "Hello world",
            }],
            loading: false,
        });
        
        const wrapper = shallow( <MultipleCustomHook />);
        expect(wrapper.find(".alerta").exists() ).toBe(false);
        // expect(wrapper.find(".mb-0").text().trim() ).toBe("Hello world");

    })
    
})