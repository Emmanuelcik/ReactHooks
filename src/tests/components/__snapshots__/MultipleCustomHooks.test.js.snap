// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the component MultipleCustomHooks should be correctly mounted 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-success"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
